### PR TITLE
ChatHeadless: hide the headless implementation and create a provider layer

### PR DESCRIPTION
**Pull request by nbramblett**
Verified tests pass and that everything builds.

Opted not to build out a provider layer in headless because there isn't anything that needs to be communicated as internal-only. The constructor is sufficient here.

However, there will be a followup PR in chat-headless-react which updates how the headless client is provided and surfaced.

### CODE REVIEW

File: packages/chat-headless/src/HeadlessProvider.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import { HeadlessConfig, ChatClient } from "./models";
+
+/**
+ * Provide an instance of {@link ChatHeadless} with all functionality built in.
+ *
+ * @public
+ */
+export function provideChatHeadless(
+  config: HeadlessConfig,
+  chatClient?: ChatClient
```
Comment (yen-tt) [inline packages/chat-headless/src/HeadlessProvider.ts:11]: during a discussion with Tom, I believe we both expressed concern for having support for two level of abstractions within one function/constructor (config that extends the core config AND a optional client to also create core). Having user pass in their own client also cause conflict with the core dependency version we have.

I think we should remove this second param. For public interface, user just need to pass in the config. We can then make a provideChatHeadlessInternal to accept the internalConfig and uses `provideChatCoreInternal`. This provide layer would create the ChatCore instance, and the ChatHeadless class's constructor will accept a ChatCore instance.

Comment (nbramblett) [reply packages/chat-headless/src/HeadlessProvider.ts]: Updated to match this proposal 👍

File: packages/chat-headless/src/ChatHeadless.ts [hunk 2]
```diff
@@ -31,12 +31,174 @@ import { getClientSdk } from "./utils/clientSdk";
...
 /**
- * Provides the functionality for interacting with a Chat Bot
- * and the relevant application state.
+ * Provides the functionality needed to interact with the Chat API in a
+ * stateful manner.
  *
  * @public
  */
-export class ChatHeadless {
+export interface ChatHeadless {
```
Comment (yen-tt) [inline packages/chat-headless/src/ChatHeadless.ts:39]: nit: move this interface into models/ folder as `ChatHeadless`, then rename this file as `ChatHeadlessIml`

File: packages/chat-headless/src/ChatHeadless.ts [hunk 3]
```diff
@@ -45,19 +207,19 @@ export class ChatHeadless {
...
+   * Constructs a new instance of the {@link ChatHeadlessImpl} class.
    *
-   * @public
+   * @internal
    *
-   * @param config - The configuration for the {@link ChatHeadless} instance
+   * @param config - The configuration for the {@link ChatHeadlessImpl} instance
    * @param chatClient - An optional override for the default {@link ChatClient} instance
    */
   constructor(config: HeadlessConfig, chatClient?: ChatClient) {
```
Comment (yen-tt) [inline packages/chat-headless/src/ChatHeadless.ts:217]: Can we update this constructor to just take in a core instance, which can be done in the provider layer, following the dependency injection approach. So:

```
//provider file
const defaultConfig: Partial<HeadlessConfig> = {
      saveToSessionStorage: true,
    };
    
provideChatHeadless(config) {
 const core: ChatCore = provideChatCore({ ...defaultConfig, ...config })
 return ChatHeadlessImpl(core)
}

provideChatHeadlessInternal(config, internalConfig) {
 const core: ChatCore = provideChatCoreInternal({ ...defaultConfig, ...config }, internalConfig)
 return ChatHeadlessImpl(core)
}

Would be easier for us to mock core as well I think

File: packages/chat-headless-react/src/ChatHeadlessProvider.tsx [hunk 4]
```diff
@@ -32,7 +32,7 @@ export function ChatHeadlessProvider(
 
   const headless = useMemo(() => {
     const configWithoutSession = { ...config, saveToSessionStorage: false };
-    const headless = new ChatHeadless(updateClientSdk(configWithoutSession));
+    const headless = provideChatHeadless(updateClientSdk(configWithoutSession));
```
Comment (yen-tt) [inline packages/chat-headless-react/src/ChatHeadlessProvider.tsx:35]: I believe we want to have ChatHeadlessProvider take in a headless instance like Search (https://github.com/yext/search-headless-react/blob/main/src/SearchHeadlessProvider.tsx). Then, the instance can be internal or public, and also can expose setter functions outside of React, which is what we want to support dynamic context in script tag

File: packages/chat-headless-react/package.json [hunk 5]
```diff
@@ -39,7 +39,7 @@
   "homepage": "https://github.com/yext/chat-headless#readme",
   "dependencies": {
     "@reduxjs/toolkit": "^1.9.5",
-    "@yext/chat-headless": "^0.5.8",
+    "@yext/chat-headless": "^0.6.0",
```
Comment (yen-tt) [inline packages/chat-headless-react/package.json:42]: i'm surprised this pass the GH action tests considering this version is not publish yet

File: packages/chat-headless-react/THIRD-PARTY-NOTICES [hunk 6]
```diff
@@ -1,707 +1,5 @@
 This file was generated with the generate-license-file npm package!
 https://www.npmjs.com/package/generate-license-file
```
Comment (yen-tt) [inline packages/chat-headless-react/THIRD-PARTY-NOTICES:2]: this looks wrong, can you try regenerating this?

File: packages/chat-headless-react/src/ChatHeadlessProvider.tsx (hunk 4 above)
Comment (nbramblett) [reply packages/chat-headless-react/src/ChatHeadlessProvider.tsx:35]: yeah I have a separate branch where I've made some changes to how headless-react works. Those will be put up separately once this is good.

File: packages/chat-headless-react/package.json (hunk 5 above)
Comment (nbramblett) [reply packages/chat-headless-react/package.json:42]: I was surprised too lol but it seems to work

File: packages/chat-headless-react/THIRD-PARTY-NOTICES (hunk 6 above)
Comment (nbramblett) [reply packages/chat-headless-react/THIRD-PARTY-NOTICES:2]: I think the issue above is breaking it. Gonna split out the headless-react changes into another pr.

File: packages/chat-headless/package.json [hunk 7]
```diff
@@ -40,7 +40,7 @@
   "dependencies": {
     "@reduxjs/toolkit": "^1.9.5",
     "@yext/analytics": "^0.6.3",
-    "@yext/chat-core": "^0.5.4"
+    "@yext/chat-core": "^0.6.0"
```
Comment (yen-tt) [inline packages/chat-headless/package.json:43]: might as well bump this to 0.6.1 for the new fix!

**Review by yen-tt — APPROVED** [review]

File: packages/chat-headless/package.json (hunk 7 above)
Comment (nbramblett) [reply packages/chat-headless/package.json:43]: oop, good call!

**Review by yen-tt — APPROVED** [review]

**Pull request merged**